# Cleansing Data

In [218]:
import pandas as pd
import re
import seaborn as sns
import matplotlib.pyplot as plt

import dataset

In [219]:
# Membaca data JSON dan mengonversinya ke DataFrame
data = pd.read_json("../../collect-data/rewind-2015/comments.json")

# Mengonversi data JSON menjadi kolom dan baris DataFrame
df = pd.json_normalize(data["comments"])

# Menampilkan lima baris pertama DataFrame
df.head()

video                      author            updated_at  like_count  \
0  fNX6Gu2F7A8                   @wibu4353  2024-04-15T09:56:11Z           0   
1  fNX6Gu2F7A8  @ginawardahtulhumairoh3722  2024-04-14T14:43:44Z           0   
2  fNX6Gu2F7A8                @alfiqri5913  2024-04-11T13:56:04Z           1   
3  fNX6Gu2F7A8          @dastananindra1566  2024-04-07T06:20:23Z           1   
4  fNX6Gu2F7A8                   @farisnst  2024-04-01T22:02:47Z           1   

                                            text  public  
0                                salam dari 2024    True  
1         Hahaha...btw niii udah 2024,GK kerasa😢    True  
2                                       the best    True  
3                            Sorry telat 9 tahun    True  
4  Ah gila udah mau 9 tahun masih apal lagu nya😂    True

Menghilangkan atribut yang tidak diperlukan

In [220]:
df = df[['author', 'updated_at', 'like_count', 'text']]
df

author            updated_at  like_count  \
0                       @wibu4353  2024-04-15T09:56:11Z           0   
1      @ginawardahtulhumairoh3722  2024-04-14T14:43:44Z           0   
2                    @alfiqri5913  2024-04-11T13:56:04Z           1   
3              @dastananindra1566  2024-04-07T06:20:23Z           1   
4                       @farisnst  2024-04-01T22:02:47Z           1   
...                           ...                   ...         ...   
28403           @GantplayBlogspot  2015-12-11T06:02:11Z           0   
28404           @ramandahafiz6293  2015-12-11T06:02:07Z           0   
28405                    @e4rtz84  2015-12-11T06:02:05Z           1   
28406      @pricillaoktaviani2376  2015-12-11T06:02:03Z           0   
28407           @abayartsgame4609  2015-12-11T06:02:01Z           1   

                                                text  
0                                    salam dari 2024  
1             Hahaha...btw niii udah 2024,GK kerasa😢  
2                                           the best  
3                                Sorry telat 9 tahun  
4      Ah gila udah mau 9 tahun masih apal lagu nya😂  
...                                              ...  
28403                                      Mantep :v  
28404                                         Boom !  
28405                        anjing canggih waaahhhh  
28406                                          firat  
28407                                              1  

[28408 rows x 4 columns]

In [221]:
df.shape

(28408, 4)

Menghapus data text yang duplikat

In [222]:
df = df.drop_duplicates(subset=['text'])

In [223]:
# melihat banyak data yang duplikat
df.duplicated().sum()
df.shape

(26480, 4)

Menghapus data author yang duplikat

In [224]:
df = df.drop_duplicates(subset=['author'])

In [225]:
# melihat banyak data yang duplikat
df.duplicated().sum()
df.shape

(24311, 4)

In [226]:
df

author            updated_at  like_count  \
0                       @wibu4353  2024-04-15T09:56:11Z           0   
1      @ginawardahtulhumairoh3722  2024-04-14T14:43:44Z           0   
2                    @alfiqri5913  2024-04-11T13:56:04Z           1   
3              @dastananindra1566  2024-04-07T06:20:23Z           1   
4                       @farisnst  2024-04-01T22:02:47Z           1   
...                           ...                   ...         ...   
28395                 @RomeuRomeu  2015-12-11T06:02:19Z           0   
28396      @pricillaoktaviani2376  2015-12-11T06:02:18Z           0   
28402           @is9production372  2015-12-11T06:02:11Z           0   
28403           @GantplayBlogspot  2015-12-11T06:02:11Z           0   
28405                    @e4rtz84  2015-12-11T06:02:05Z           1   

                                                text  
0                                    salam dari 2024  
1             Hahaha...btw niii udah 2024,GK kerasa😢  
2                                           the best  
3                                Sorry telat 9 tahun  
4      Ah gila udah mau 9 tahun masih apal lagu nya😂  
...                                              ...  
28395                                          woeee  
28396                                              1  
28402                                    =))) kerenn  
28403                                      Mantep :v  
28405                        anjing canggih waaahhhh  

[24311 rows x 4 columns]

Menghapus nilai yang kosong

In [227]:
df = df.dropna()

In [228]:
# melihat data yang kosong
df.isnull().sum()

author        0
updated_at    0
like_count    0
text          0
dtype: int64

Melihat banyak data setelah cleansing

In [229]:
df.shape

(24311, 4)

In [230]:
df

author            updated_at  like_count  \
0                       @wibu4353  2024-04-15T09:56:11Z           0   
1      @ginawardahtulhumairoh3722  2024-04-14T14:43:44Z           0   
2                    @alfiqri5913  2024-04-11T13:56:04Z           1   
3              @dastananindra1566  2024-04-07T06:20:23Z           1   
4                       @farisnst  2024-04-01T22:02:47Z           1   
...                           ...                   ...         ...   
28395                 @RomeuRomeu  2015-12-11T06:02:19Z           0   
28396      @pricillaoktaviani2376  2015-12-11T06:02:18Z           0   
28402           @is9production372  2015-12-11T06:02:11Z           0   
28403           @GantplayBlogspot  2015-12-11T06:02:11Z           0   
28405                    @e4rtz84  2015-12-11T06:02:05Z           1   

                                                text  
0                                    salam dari 2024  
1             Hahaha...btw niii udah 2024,GK kerasa😢  
2                                           the best  
3                                Sorry telat 9 tahun  
4      Ah gila udah mau 9 tahun masih apal lagu nya😂  
...                                              ...  
28395                                          woeee  
28396                                              1  
28402                                    =))) kerenn  
28403                                      Mantep :v  
28405                        anjing canggih waaahhhh  

[24311 rows x 4 columns]

Dari yang awalnya berjumlah 28408 row berkurang menjadi 24311 row

Mengubah data text menjadi lowercase

In [231]:
df['text'] = df['text'].str.lower()

In [232]:
df

author            updated_at  like_count  \
0                       @wibu4353  2024-04-15T09:56:11Z           0   
1      @ginawardahtulhumairoh3722  2024-04-14T14:43:44Z           0   
2                    @alfiqri5913  2024-04-11T13:56:04Z           1   
3              @dastananindra1566  2024-04-07T06:20:23Z           1   
4                       @farisnst  2024-04-01T22:02:47Z           1   
...                           ...                   ...         ...   
28395                 @RomeuRomeu  2015-12-11T06:02:19Z           0   
28396      @pricillaoktaviani2376  2015-12-11T06:02:18Z           0   
28402           @is9production372  2015-12-11T06:02:11Z           0   
28403           @GantplayBlogspot  2015-12-11T06:02:11Z           0   
28405                    @e4rtz84  2015-12-11T06:02:05Z           1   

                                                text  
0                                    salam dari 2024  
1             hahaha...btw niii udah 2024,gk kerasa😢  
2                                           the best  
3                                sorry telat 9 tahun  
4      ah gila udah mau 9 tahun masih apal lagu nya😂  
...                                              ...  
28395                                          woeee  
28396                                              1  
28402                                    =))) kerenn  
28403                                      mantep :v  
28405                        anjing canggih waaahhhh  

[24311 rows x 4 columns]

Menghapus data dengan atribut text null

In [233]:
df = df[df['text'] != '']

In [234]:
df

author            updated_at  like_count  \
0                       @wibu4353  2024-04-15T09:56:11Z           0   
1      @ginawardahtulhumairoh3722  2024-04-14T14:43:44Z           0   
2                    @alfiqri5913  2024-04-11T13:56:04Z           1   
3              @dastananindra1566  2024-04-07T06:20:23Z           1   
4                       @farisnst  2024-04-01T22:02:47Z           1   
...                           ...                   ...         ...   
28395                 @RomeuRomeu  2015-12-11T06:02:19Z           0   
28396      @pricillaoktaviani2376  2015-12-11T06:02:18Z           0   
28402           @is9production372  2015-12-11T06:02:11Z           0   
28403           @GantplayBlogspot  2015-12-11T06:02:11Z           0   
28405                    @e4rtz84  2015-12-11T06:02:05Z           1   

                                                text  
0                                    salam dari 2024  
1             hahaha...btw niii udah 2024,gk kerasa😢  
2                                           the best  
3                                sorry telat 9 tahun  
4      ah gila udah mau 9 tahun masih apal lagu nya😂  
...                                              ...  
28395                                          woeee  
28396                                              1  
28402                                    =))) kerenn  
28403                                      mantep :v  
28405                        anjing canggih waaahhhh  

[24310 rows x 4 columns]

Menghapus baris dengan nilai yang hanya berupa angka pada kolom 'text'

In [235]:
df = df[~df['text'].str.isnumeric()]

In [236]:
df

author            updated_at  like_count  \
0                       @wibu4353  2024-04-15T09:56:11Z           0   
1      @ginawardahtulhumairoh3722  2024-04-14T14:43:44Z           0   
2                    @alfiqri5913  2024-04-11T13:56:04Z           1   
3              @dastananindra1566  2024-04-07T06:20:23Z           1   
4                       @farisnst  2024-04-01T22:02:47Z           1   
...                           ...                   ...         ...   
28394            @daffareyhan7744  2015-12-11T06:02:19Z           0   
28395                 @RomeuRomeu  2015-12-11T06:02:19Z           0   
28402           @is9production372  2015-12-11T06:02:11Z           0   
28403           @GantplayBlogspot  2015-12-11T06:02:11Z           0   
28405                    @e4rtz84  2015-12-11T06:02:05Z           1   

                                                text  
0                                    salam dari 2024  
1             hahaha...btw niii udah 2024,gk kerasa😢  
2                                           the best  
3                                sorry telat 9 tahun  
4      ah gila udah mau 9 tahun masih apal lagu nya😂  
...                                              ...  
28394                                         first?  
28395                                          woeee  
28402                                    =))) kerenn  
28403                                      mantep :v  
28405                        anjing canggih waaahhhh  

[24287 rows x 4 columns]

Menghapus baris dengan nilai yang hanya memiliki 1 karakter pada kolom 'text'

In [237]:
df = df[df['text'].str.len() > 1]

In [238]:
df

author            updated_at  like_count  \
0                       @wibu4353  2024-04-15T09:56:11Z           0   
1      @ginawardahtulhumairoh3722  2024-04-14T14:43:44Z           0   
2                    @alfiqri5913  2024-04-11T13:56:04Z           1   
3              @dastananindra1566  2024-04-07T06:20:23Z           1   
4                       @farisnst  2024-04-01T22:02:47Z           1   
...                           ...                   ...         ...   
28394            @daffareyhan7744  2015-12-11T06:02:19Z           0   
28395                 @RomeuRomeu  2015-12-11T06:02:19Z           0   
28402           @is9production372  2015-12-11T06:02:11Z           0   
28403           @GantplayBlogspot  2015-12-11T06:02:11Z           0   
28405                    @e4rtz84  2015-12-11T06:02:05Z           1   

                                                text  
0                                    salam dari 2024  
1             hahaha...btw niii udah 2024,gk kerasa😢  
2                                           the best  
3                                sorry telat 9 tahun  
4      ah gila udah mau 9 tahun masih apal lagu nya😂  
...                                              ...  
28394                                         first?  
28395                                          woeee  
28402                                    =))) kerenn  
28403                                      mantep :v  
28405                        anjing canggih waaahhhh  

[24256 rows x 4 columns]

Fungsi untuk menghapus karakter tunggal yang terpisah dari kata-kata

In [239]:
def remove_single_characters(text):
    # Mencari dan menghapus karakter tunggal yang terpisah dari kata-kata
    return re.sub(r'\b\w\b', '', text)

In [240]:
# Menghapus karakter tunggal yang terpisah dari nilai pada kolom 'text'
# Menggunakan .loc untuk mengakses DataFrame
df.loc[:, 'text'] = df['text'].apply(remove_single_characters)

In [241]:
df

author            updated_at  like_count  \
0                       @wibu4353  2024-04-15T09:56:11Z           0   
1      @ginawardahtulhumairoh3722  2024-04-14T14:43:44Z           0   
2                    @alfiqri5913  2024-04-11T13:56:04Z           1   
3              @dastananindra1566  2024-04-07T06:20:23Z           1   
4                       @farisnst  2024-04-01T22:02:47Z           1   
...                           ...                   ...         ...   
28394            @daffareyhan7744  2015-12-11T06:02:19Z           0   
28395                 @RomeuRomeu  2015-12-11T06:02:19Z           0   
28402           @is9production372  2015-12-11T06:02:11Z           0   
28403           @GantplayBlogspot  2015-12-11T06:02:11Z           0   
28405                    @e4rtz84  2015-12-11T06:02:05Z           1   

                                               text  
0                                   salam dari 2024  
1            hahaha...btw niii udah 2024,gk kerasa😢  
2                                          the best  
3                                sorry telat  tahun  
4      ah gila udah mau  tahun masih apal lagu nya😂  
...                                             ...  
28394                                        first?  
28395                                         woeee  
28402                                   =))) kerenn  
28403                                      mantep :  
28405                       anjing canggih waaahhhh  

[24256 rows x 4 columns]

Fungsi untuk mengubah kata dengan huruf homogen yang lebih dari 1 menjadi 1 huruf saja

In [242]:
def remove_repeated_letters(text):
    return re.sub(r'(\w)\1+', r'\1', text)

In [243]:
# Mengubah data dalam DataFrame menggunakan fungsi remove_repeated_letters
df.loc[:, 'text'] = df['text'].apply(remove_repeated_letters)
df

author            updated_at  like_count  \
0                       @wibu4353  2024-04-15T09:56:11Z           0   
1      @ginawardahtulhumairoh3722  2024-04-14T14:43:44Z           0   
2                    @alfiqri5913  2024-04-11T13:56:04Z           1   
3              @dastananindra1566  2024-04-07T06:20:23Z           1   
4                       @farisnst  2024-04-01T22:02:47Z           1   
...                           ...                   ...         ...   
28394            @daffareyhan7744  2015-12-11T06:02:19Z           0   
28395                 @RomeuRomeu  2015-12-11T06:02:19Z           0   
28402           @is9production372  2015-12-11T06:02:11Z           0   
28403           @GantplayBlogspot  2015-12-11T06:02:11Z           0   
28405                    @e4rtz84  2015-12-11T06:02:05Z           1   

                                               text  
0                                   salam dari 2024  
1              hahaha...btw ni udah 2024,gk kerasa😢  
2                                          the best  
3                                 sory telat  tahun  
4      ah gila udah mau  tahun masih apal lagu nya😂  
...                                             ...  
28394                                        first?  
28395                                           woe  
28402                                    =))) keren  
28403                                      mantep :  
28405                             anjing cangih wah  

[24256 rows x 4 columns]

Memeriksa panjang teks pada kolom 'text' dan menyimpan hanya baris dengan panjang minimal 20 karakter

In [244]:
df = df[(df['text'].str.len() >= 20) & (df['text'].str.len() <= 200)]
df

author            updated_at  like_count  \
1      @ginawardahtulhumairoh3722  2024-04-14T14:43:44Z           0   
4                       @farisnst  2024-04-01T22:02:47Z           1   
5                          @Pziuw  2024-04-01T15:17:42Z           2   
7                   @Deviii_al_al  2024-03-29T00:35:25Z           3   
9                    @rekyrms9452  2024-03-25T03:55:30Z           1   
...                           ...                   ...         ...   
28334     @yuandaristagayatri4280  2015-12-11T06:03:39Z           0   
28351          @muhammadyasin5544  2015-12-11T06:03:07Z           0   
28356         @ichsanramadhan1832  2015-12-11T06:02:59Z           0   
28361           @sesariatitis3625  2015-12-11T06:02:50Z           0   
28366                   @Rapipuuu  2015-12-11T06:02:46Z           0   

                                                    text  
1                   hahaha...btw ni udah 2024,gk kerasa😢  
4           ah gila udah mau  tahun masih apal lagu nya😂  
5      zaman di mna jdi youtuber hrus dgn karya terba...  
7      nyesek bgt wlapun ak di situ msih tk masih suk...  
9      abis liat yg 2023, btw bryan furan yg bikin gw...  
...                                                  ...  
28334  keren 😃 njir videonya bikin ane merinding : te...  
28351          akhirnya youtube rewind indonesia 2015!!!  
28356                             gw ke berapa kampret :  
28361                               likers ke 62!!!!!!!!  
28366  congratulation guys for make this awesome vide...  

[17764 rows x 4 columns]

Fungsi untuk membersihkan teks dari beberapa pola yang umum ditemukan dalam data teks

In [245]:
def clean_data_text(text):
    # Menghapus username Twitter
    text = re.sub(r'@[A-Za-z0-9_]+', '', text)
    # Menghapus hashtag
    text = re.sub(r'#\w+', '', text)
    # Menghapus 'RT' (retweet)
    text = re.sub(r'RT[\s]+', '', text)
    # Menghapus URL
    text = re.sub(r'https?://\S+', '', text)
    # Menghapus karakter non-alfanumerik
    text = re.sub(r'[^A-Za-z0-9 ]', '', text)
    # Menghapus spasi ekstra di awal dan akhir teks, serta spasi berlebih di antara kata
    text = re.sub(r'\s+', ' ', text).strip()
    return text

In [246]:
df

author            updated_at  like_count  \
1      @ginawardahtulhumairoh3722  2024-04-14T14:43:44Z           0   
4                       @farisnst  2024-04-01T22:02:47Z           1   
5                          @Pziuw  2024-04-01T15:17:42Z           2   
7                   @Deviii_al_al  2024-03-29T00:35:25Z           3   
9                    @rekyrms9452  2024-03-25T03:55:30Z           1   
...                           ...                   ...         ...   
28334     @yuandaristagayatri4280  2015-12-11T06:03:39Z           0   
28351          @muhammadyasin5544  2015-12-11T06:03:07Z           0   
28356         @ichsanramadhan1832  2015-12-11T06:02:59Z           0   
28361           @sesariatitis3625  2015-12-11T06:02:50Z           0   
28366                   @Rapipuuu  2015-12-11T06:02:46Z           0   

                                                    text  
1                   hahaha...btw ni udah 2024,gk kerasa😢  
4           ah gila udah mau  tahun masih apal lagu nya😂  
5      zaman di mna jdi youtuber hrus dgn karya terba...  
7      nyesek bgt wlapun ak di situ msih tk masih suk...  
9      abis liat yg 2023, btw bryan furan yg bikin gw...  
...                                                  ...  
28334  keren 😃 njir videonya bikin ane merinding : te...  
28351          akhirnya youtube rewind indonesia 2015!!!  
28356                             gw ke berapa kampret :  
28361                               likers ke 62!!!!!!!!  
28366  congratulation guys for make this awesome vide...  

[17764 rows x 4 columns]

In [249]:
df.loc[:, 'text'] = df['text'].apply(clean_data_text)
df

author            updated_at  like_count  \
1      @ginawardahtulhumairoh3722  2024-04-14T14:43:44Z           0   
4                       @farisnst  2024-04-01T22:02:47Z           1   
5                          @Pziuw  2024-04-01T15:17:42Z           2   
7                   @Deviii_al_al  2024-03-29T00:35:25Z           3   
9                    @rekyrms9452  2024-03-25T03:55:30Z           1   
...                           ...                   ...         ...   
28334     @yuandaristagayatri4280  2015-12-11T06:03:39Z           0   
28351          @muhammadyasin5544  2015-12-11T06:03:07Z           0   
28356         @ichsanramadhan1832  2015-12-11T06:02:59Z           0   
28361           @sesariatitis3625  2015-12-11T06:02:50Z           0   
28366                   @Rapipuuu  2015-12-11T06:02:46Z           0   

                                                    text  
1                        hahahabtw ni udah 2024gk kerasa  
4             ah gila udah mau tahun masih apal lagu nya  
5      zaman di mna jdi youtuber hrus dgn karya terba...  
7      nyesek bgt wlapun ak di situ msih tk masih suk...  
9      abis liat yg 2023 btw bryan furan yg bikin gw ...  
...                                                  ...  
28334  keren njir videonya bikin ane merinding terus ...  
28351             akhirnya youtube rewind indonesia 2015  
28356                               gw ke berapa kampret  
28361                                       likers ke 62  
28366  congratulation guys for make this awesome vide...  

[17764 rows x 4 columns]

Simpan clean data dalam file json

In [250]:
import json

# Membuat list of dictionaries dengan key yang diinginkan
comments_data = []
for index, row in df.iterrows():
    comment_dict = {
        'author': row['author'],
        'updated_at': row['updated_at'],
        'like_count': row['like_count'],
        'text': row['text']
    }
    comments_data.append(comment_dict)

# Membuat dictionary yang berisi key 'clean_comments' dengan nilai dari comments_data
data = {'clean_comments': comments_data}

# Menyimpan dictionary ke dalam file JSON
with open('clean_comments.json', 'w') as file:
    json.dump(data, file, indent=4)